### PR TITLE
[CLICKHOUSE-3476] add invalidate_query for ClickHouse in Dictionary

### PR DESCRIPTION
**Pull request by VadimPlh**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/src/Dictionaries/ClickHouseDictionarySource.cpp [hunk 1]
```diff
@@ -143,4 +159,19 @@ BlockInputStreamPtr ClickHouseDictionarySource::createStreamForSelectiveLoad(con
...
+{
+    Block invalidate_sample_block;
+
+    if (is_local)
+    {
+        auto input_block = executeQuery(request, context, true).in;
+        return readInvalidateQuery(dynamic_cast<IProfilingBlockInputStream&>((*input_block)));
+    }
+ 
+    auto invalidate_stream = RemoteBlockInputStream(pool, request, invalidate_sample_block, context);
```
Comment (alesapin) [inline dbms/src/Dictionaries/ClickHouseDictionarySource.cpp]: Why not `RemoteBlockInputStream invalidate_stream(pool, request, invalidate_sample_block, context);`?

File: dbms/src/Dictionaries/ClickHouseDictionarySource.cpp [hunk 2]
```diff
@@ -143,4 +159,19 @@ BlockInputStreamPtr ClickHouseDictionarySource::createStreamForSelectiveLoad(con
     return std::make_shared<RemoteBlockInputStream>(pool, query, sample_block, context);
 }
 
+std::string ClickHouseDictionarySource::doInvalidateQuery(const std::string & request) const
+{
+    Block invalidate_sample_block;
```
Comment (alesapin) [inline dbms/src/Dictionaries/ClickHouseDictionarySource.cpp]: Unused variable in case of `is_local`. Better to create `if (is_local) ... else ...`.

Comment (alesapin) [inline dbms/src/Dictionaries/ClickHouseDictionarySource.cpp]: This block is `sample_block`, which is used as header. It must contain single string column. Look at similar dictionaries: https://github.com/yandex/ClickHouse/blob/master/dbms/src/Dictionaries/MySQLDictionarySource.cpp#L219-L221 and https://github.com/yandex/ClickHouse/blob/master/dbms/src/Dictionaries/ODBCDictionarySource.cpp#L187-L189

**Review by alesapin — CHANGES_REQUESTED** [review]
Need test here https://github.com/yandex/ClickHouse/tree/master/dbms/tests/external_dictionaries or here https://github.com/yandex/ClickHouse/tree/master/dbms/tests/integration.

File: dbms/src/Dictionaries/ClickHouseDictionarySource.cpp (hunk 1 above)
Comment (VadimPE) [reply dbms/src/Dictionaries/ClickHouseDictionarySource.cpp]: For my fun)

File: dbms/src/Dictionaries/ClickHouseDictionarySource.cpp (hunk 2 above)
Comment (VadimPE) [reply dbms/src/Dictionaries/ClickHouseDictionarySource.cpp]: No! It does not work with CH.

Comment (alesapin) [reply dbms/src/Dictionaries/ClickHouseDictionarySource.cpp]: ok

**Comment by alexey-milovidov** [comment]
Ok. But test is missing.

**Pull request merged**